### PR TITLE
add support for decoding vc-1 interlaced content

### PR DESCRIPTION
**Pull request by carpalis**
These commits add support for hardware decoding vc-1 interlaced content. They start off with some cleaning up of the code followed by the real implementation. The code has only been tested on gen75 hardware. This code needs 01org/libva#138 to compile. The changes necessary to make FFmpeg hardware decode interlaced vc-1 content can be found in my repository: https://github.com/carpalis/FFmpeg/tree/vc1interlace

Signed-off-by: Jerome Borsboom <jerome.borsboom@carpalis.nl>

### CODE REVIEW

**Comment by jayesht** [comment]
Hi @carpalis, could you please suggest a few tests to validate this PR? 

I patched intel-vaapi-driver with this PR and used https://github.com/carpalis/FFmpeg/tree/vc1interlace to exercise VAAPI and this driver, but do not see difference while decoding VC1 interlaced clip with or without this PR, using ffplay.

**Comment by carpalis** [comment]
I am not an ffmpeg wizard, but AFAIK ffplay does not use hardware to decode VC-1. What works for me is:

ffmpeg -hwaccel vaapi -i {in-filename} {out-filename}.yuv

This decodes the input stream to a sequence of raw NV12 surfaces. This sequence can be played with yuvplayer which can be found on SourceForge. Please be aware that the resulting file can become very large.

The difference in decoding speed between software and hardware decoding can be checked with:

ffmpeg -i {in-filename} -f null /dev/null
ffmpeg -hwaccel vaapi -i {in-filename} -f null /dev/null

Another possibility is to use Kodi. No patches are necessary for Kodi. Only the ffmpeg, va and vaapi libraries need patching.

I suppose that VLC can be made to work, but I have not check that myself.

File: src/gen75_mfd.c [hunk 1]
```diff
@@ -1525,7 +1524,7 @@ gen75_mfd_init_vc1_surface(VADriverContextP ctx,
     if (gen7_vc1_surface->dmv == NULL) {
         gen7_vc1_surface->dmv = dri_bo_alloc(i965->intel.bufmgr,
                                              "direct mv w/r buffer",
-                                             width_in_mbs * height_in_mbs * 64,
+                                             128 * height_in_mbs * 64,
```
Comment (jayesht) [inline src/gen75_mfd.c]: Hardware assumption regarding width_in_mbs=128 can be documented as comment.

File: src/gen7_mfd.c [hunk 2]
```diff
@@ -1260,7 +1259,7 @@ gen7_mfd_init_vc1_surface(VADriverContextP ctx,
     if (gen7_vc1_surface->dmv == NULL) {
         gen7_vc1_surface->dmv = dri_bo_alloc(i965->intel.bufmgr,
                                              "direct mv w/r buffer",
-                                             width_in_mbs * height_in_mbs * 64,
+                                             128 * height_in_mbs * 64,
```
Comment (jayesht) [inline src/gen7_mfd.c]: Hardware assumption regarding width_in_mbs=128 can be documented as comment.

**Comment by jayesht** [comment]
[vc1_hw_sw_decode.log](https://github.com/01org/intel-vaapi-driver/files/1537494/vc1_hw_sw_decode.log)

a) Patchsets look good to me. No additional review comments.
b) Attached is software vs hardware decoding via patched ffmpeg for an interlaced vc-1 clip. The time taken for hardware decode is more.
c) ffplay back of the resultant .yuv's via software and hardware decode is different. Hardware decoded yuv lost its color information.

The above observations are with ToT libva, patched intel-vaapi-driver with this PR on Kabylake platform.
[ilaced_twomv.zip](https://github.com/01org/intel-vaapi-driver/files/1537499/ilaced_twomv.zip)

**Comment by carpalis** [comment]
Hi jayesht,

a) Thank you for reviewing my patch set. I have already included some comment lines as you proposed.

b) I am a bit baffled by this one. From your remark under c), I assume that hardware decoding produces a recognizable image. But taking 11 seconds to decode 13 frames is horrible. Could you do something like this to see where the 11 seconds are being eaten?

ffmpeg -loglevel debug -hwaccel vaapi -i ilaced_twomv.vc1 -f null /dev/null 2>&1 | ts "%H:%M:%.S"

And just to make sure, could you check with dmesg that there are no GPU hangs when hardware decoding?

When I decode with fedora 25 on Haswell, using vaapi lib 1.8.3/0.40.0 and ffmpeg 3.1.11 I get these numbers:

HW decode:
real    0m0.378s
user    0m0.104s
sys     0m0.058s

and 

SW decode:
real    0m0.371s
user    0m0.343s
sys     0m0.023s.

c) Please make sure that you force the pix_fmt to nv12 when playing hardware decoded .yuv's. Software decoding delivers .yuv's in yuv420p pix_fmt which ffplay eats without any additional switches.

**Comment by jayesht** [comment]
b) I played the clip with  2>&1 | ts "%H:%M:%.S". Attached is log output. MapSurface is consuming the time excessively. I see gpu hang too with this clip in dmesg.

          b.1) I tried other clip from : http://samples.mplayerhq.hu/V-codecs/WVC1/vc1-interlaced-bframes.m2ts and dont see gpu hang.  Also, time taken is comparable between HW decode and SW decode. ~12sec for HW and ~14sec for SW.

c)  I supplied yuv420p and nv12 to pix_fmt - but ffplay still strips down color information. 
In case of : http://samples.mplayerhq.hu/V-codecs/WVC1/vc1-interlaced-bframes.m2ts clip, resultant yuv playback from ffmpeg hardware decode is pixelated and devoide of color information for first few frames.
[vc1_hw.txt](https://github.com/01org/intel-vaapi-driver/files/1540314/vc1_hw.txt)

**Comment by carpalis** [comment]
Are you sure your ffmpeg is patched with the last three commits in my ffmpeg/vc1interlace repository? The bold lines below should not be there. That frame is a field interlaced P/P-frame. P-fields do not have bitplanes and, therefore, no Param buffer with type 2 should be sent to the vaapi library.

09:45:34.133514 [vc1 @ 0x2c75500] VOP DQuant info
09:45:34.133717 [vc1 @ 0x2c75500] Param buffer (type 0, 104 bytes) is 0x8000003.
**09:45:34.133754 [vc1 @ 0x2c75500] Param buffer (type 2, 4080 bytes) is 0x8000002.**
09:45:34.133788 [vc1 @ 0x2c75500] Slice 0 param buffer (36 bytes) is 0x8000001.
09:45:34.133820 [vc1 @ 0x2c75500] Slice 0 data buffer (161575 bytes) is 0x8000000.
09:45:34.133869 [vc1 @ 0x2c75500] Decode to surface 0x4000004.
09:45:34.134901 [vc1 @ 0x2c75500] VOP DQuant info
09:45:34.134939 [vc1 @ 0x2c75500] Param buffer (type 0, 104 bytes) is 0x8000000.
**09:45:34.134955 [vc1 @ 0x2c75500] Param buffer (type 2, 4080 bytes) is 0x8000001.**
09:45:34.134971 [vc1 @ 0x2c75500] Slice 0 param buffer (36 bytes) is 0x8000002.
09:45:34.134992 [vc1 @ 0x2c75500] Slice 0 data buffer (138987 bytes) is 0x8000003.
09:45:34.135008 [vc1 @ 0x2c75500] Decode to surface 0x4000004.
09:45:34.135264 [AVHWFramesContext @ 0x2c7f260] Map surface 0x4000005.
09:45:38.539672 [AVHWFramesContext @ 0x2c7f260] Unmap surface 0x4000005.

**Comment by jayesht** [comment]
My bad. I was on master branch on your fork of ffmpeg.

I moved to interlace branch and this patch set stands in great shape.

LGTM and approved. @xhaihao

**Comment by xhaihao** [comment]
@jayesht could you change the review status to approved?

**Pull request merged**